eth/client.go

```diff
@@ -738,8 +739,109 @@ func (c *client) Vote(pollAddr ethcommon.Address, choiceID *big.Int) (*types.Tra
 	return poll.Vote(opts, choiceID)
 }
 
+func (c *client) Reward() (*types.Transaction, error) {
+	addr := c.accountManager.Account().Address
+
+	t, err := c.GetTranscoder(addr)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to get transcoder %v", addr.Hex())
+	}
+
+	mintable, err := c.CurrentMintableTokens()
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to get curernt mintable tokens")
+	}
+
+	totalBonded, err := c.GetTotalBonded()
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to get total bonded")
+	}
+
+	// reward = (current mintable tokens for the round * active transcoder stake) / total active stake
+	reward := new(big.Int).Div(new(big.Int).Mul(mintable, t.DelegatedStake), totalBonded)
+
+	hints, err := c.simulateTranscoderPoolUpdate(addr, reward.Add(reward, t.DelegatedStake))
+	if err != nil {
+		glog.Errorf("Unable to calculate hints, submitting reward transaction without hints. err=%v", err)
+	}
+	return c.RewardWithHint(hints.PosPrev, hints.PosNext)
+
+}
+
 // Helpers
 
+// simulateTranscoderPoolUpdate simulated an  update to the transcoder pool and returns the positional hints for a delegator accordingly.
+// if the delegator will not be in the updated set 'nil' will be returned.
+func (c *client) simulateTranscoderPoolUpdate(del ethcommon.Address, newStake *big.Int) (lpTypes.StakingHints, error) {
+
+	// get max pool size
+	maxSize, err := c.GetTranscoderPoolMaxSize()
+	if err != nil {
+		return lpTypes.StakingHints{}, err
+	}
+	maxSizeInt := int(maxSize.Int64())
+
+	// get the transcoder  pool
+	transcoders, err := c.TranscoderPool()
+	if err != nil {
+		return lpTypes.StakingHints{}, err
+	}
+
+	// if the transcoder pool is full and the newStake for 'del' is lower  than that of the last transcoders
+	// 'del' will not be in the active set and thus there are no hints to return.
+	if len(transcoders) == maxSizeInt && newStake.Cmp(transcoders[maxSizeInt-1].DelegatedStake) <= 0 {
+		return lpTypes.StakingHints{}, nil
+	}
+
+	// Find the transcoder and update it's stake
```

Comment:
Would we be able to test more of this logic if it was moved into the pure function `findTranscoderHints()` and the values returned by the ETH calls in this method passed into the pure function?

Reply:
We could but it seemed a little spaghetti-ish 🍝  to me so I opted for the in-between solution.

Reply:
What about something like the following:

```golang
func (c *client) Reward() (*types.Transaction, error) {
	...

	transcoders, err := c.TranscoderPool()
	if err != nil {
		return nil, err
	}

	hint := simulateTranscoderPoolUpdate(addr, reward.Add(reward, ep.TotalStake), transcoders)

	return c.RewardWithHint(hint.PosPrev, hint.PosNext)
}

func simulateTranscoderPoolJoin(del ethcommon.Address, newStake *big.Int, transcoders []*lpTypes.Transcoder, maxSize int) lpTypes.StakingHints {
	if len(transcoders) == maxSize {
		if newStake.Cmp(transcoders[len(transcoders)-1].DelegatedStake) <= 0 {
			return lpTypes.StakingHints{}
		}

		// Evict the last transcoder
		transcoders = transcoders[1:]
	}

	return simulateTranscoderPoolUpdate(del, newStake, transcoders)
}

func simulateTranscoderPoolUpdate(del ethcommon.Address, newStake *big.Int, transcoders []*lpTypes.Transcoder) lpTypes.StakingHints {
	// Find the transcoder and update it's stake
	for i, t := range transcoders {
		// if 'del' is in the pool simply update it's stake
		if t.Address == del {
			t.DelegatedStake = newStake
		} else if i == len(transcoders)-1 {
			// 'del' is not in the pool, insert it and sort later
			transcoders = append(transcoders, &lpTypes.Transcoder{
				Address:        del,
				DelegatedStake: newStake,
			})
		}
	}

	// re-sort the list after the stake update
	sort.SliceStable(transcoders, func(i, j int) bool {
		return transcoders[i].DelegatedStake.Cmp(transcoders[j].DelegatedStake) > 0
	})

	hints := lpTypes.StakingHints{}

	// do a linear search to get the previous and next transcoder relative to 'del'
	for i, t := range transcoders {
		if t.Address == del {
			if i == 0 && len(transcoders) > 1 {
				// 'del' is head
				hints.PosNext = transcoders[i+1].Address
			} else if i == len(transcoders)-1 && len(transcoders) > 1 {
				// 'del' is tail
				hints.PosPrev = transcoders[i-1].Address
			} else {
				hints.PosNext = transcoders[i+1].Address
				hints.PosPrev = transcoders[i-1].Address
			}
		}
	}

	return hints
}
```

The pool update simulation logic in `simulateTranscoderPoolUpdate()` could be unit tested - a convenient way to write the tests might be to use the [rapid](https://github.com/flyingmutant/rapid) package to auto-generate test cases that can be executed to not only ensure that the relevant transcoder's stake is updated in the pool, but to also ensure that an invariant (i.e. the transcoder pool slice is sorted after each function call) is preserved. The rapid package is used in a [few other places](https://github.com/livepeer/go-livepeer/search?q=rapid&unscoped_q=rapid) in the codebase.

The eviction logic could be isolated in `simulateTranscoderPoolJoin()` which calls `simulateTranscoderPoolUpdate()` internally after executing the eviction logic. This logic could be unit tested too. This might be helpful since the eviction logic only matters when `bondWithHint()` is invoked in the future.

Reply:
I don't really like having a separate method for bonding and it takes curves around the bigger issue at hand which is the testability of our ETH client package, not that I think the latter should be in scope for this PR either. 

I know this reply doesn't get us farther ahead but do you strongly feel that the current testability of this new logic is insufficient? 

To me it feels like being able to correct the right hints given a transcoder and sorted transcoder list is at the core here especially if we get rid of L796.

Reply:
> I don't really like having a separate method for bonding and it takes curves around the bigger issue at hand which is the testability of our ETH client package

I think there are two points here:

1. Pushing as much logic as possible into a pure function so that it can be tested
2. Using a separate function when calculating the hint for `bondWithHint()`

Regarding 1: Suppose we had the ability to stub the contract bindings right now. Presumably, we'd be writing tests that stub the calls to the contract bindings and then execute the code paths after those calls. If that is the case, why not write the tests for those code paths now? Then, if we are able to stub the contract bindings in the future we would already have the tests for the other code paths and would only need to add tests that involve the stub contract binding calls.

Regarding 2: The main reason I suggested this is to avoid adding code that is never tested either via a test suite or manually while still allowing for the possibility of adding that code easily in the future. At the moment, we're including eviction logic that is not unit testable and not manually testable since it is not triggered anywhere (since it is not required for `rewardWithHint()`). One way to address this would be to add the eviction logic later in a function that invokes `simulateTranscoderPoolUpdate()` under the hood when we actually need to invoke that logic. 

What is the downside of having a separate function for bonding? Seem reasonable given that joining the transcoder pool can be considered an update to the transcoder pool preceded by eviction logic that is only necessary for the join operation.

> To me it feels like being able to correctly calculate the hints given a transcoder and sorted transcoder list is at the core

Yeah those operations are important, but they only work if the logic for finding the right position for the transcoder is also correct.

Reply:
Okay I understand your points.

To me it just felt a little bit excessive considering we don't have testing for the rest of the `Client` methods and considering there's actually no real grave negative side effects possible. 

But if you feel we should tackle full testability now , so be it happy to make the change.

Reply:
> we don't have testing for the rest of the Client methods

If the logic introduced in this PR was simpler (i.e. just a couple lines of a code that use values from contract calls) then I would be more inclined to not add tests. But, the reasons why I do think it makes sense to add additional tests for this PR are:

- The logic for correctly calculating a hint is more involved
- Most of the logic required to correctly calculate a hint does not depend on contract calls (besides using initial values from contract calls)

Lastly, I think adding a few tests actually makes our lives easier. They can confirm for us that the hint is actually correctly calculated. If we were only manually testing, it would be more difficult to determine whether the hint is correctly calculated (we'd see a gas cost reduction, but we might not be certain that this is the greatest gas cost reduction that is possible).

If we want to minimize some of the testing, I'd also be ok with omitting the proper eviction logic until we actually need it i.e. when we work on #1567.

Reply:
I think now we have 

- a single method
- full test coverage
- ejections and joins covered
- changing delegates covered (two calls required, one for the old del and one for the new del)

There's been a quite a few changes so reviewing the final diff might be more helpful.